### PR TITLE
Home assistant integration

### PR DESCRIPTION
**Pull request by foxy82**
### Description

Home Assistant integration - this will connect to Home Assistant Web Socket API and exposes lights to diy hue. This means you can then control your Home Assistant lights in the Diy Hue App (and hopefully the normal Hue App - to be tested).

To use this there is a homeassistant section in config.json:

```yaml
    "homeassistant": {
      "enabled": false,
      "homeAssistantIp": "127.0.0.1", 
      "homeAssistantPort": 8123, 
      "homeAssistantToken": "",
      "homeAssistantIncludeByDefault": true
    }
```
Hopefully most of the settings are self explanatory. The `homeAssistantToken` is a long lived access token - see here to generate: https://developers.home-assistant.io/docs/auth_api/#long-lived-access-token

`homeAssistantIncludeByDefault` - when set to true all "light.*" except those configured to be excluded (see below) will be added. If set to false then no lights will be added unless they are specifically included. (see below)

### Other diyhue changes incorporated into this change

There are a few bug fixes/changes made to diyhue as part of this PR

- Fixed concurrent modification of the lights map if sync lights and discover are running at the same time.
- Fixed issue where a call to /api/.../sensors/new was returning the list of lights which broke the Hue App 
- I've made all the protocols configurable but left default as enabled to maintain backward compatible behaviour. The reason for this is that if someone is using the Home Assistant integration then the chances are all things like Sonoff, Shelly, ESPHome are all in Home Assistant and so it allows the user to only use the ones in Home Assistant without having lights appear twice or using a mix and match set of services - however the user can configure how they want.

### Include / Exclude Lights from Home Assistant

To include or exclude individual lights you do this in Home Assistant by customizing the entity. 

You can do this in several ways - one way is to use the UI using the customize entities page : https://www.home-assistant.io/docs/configuration/customizing-devices/#customization-using-the-ui

You need to add an attribute called `diyhue` and set it to `include` to include the light and `exclude` to exclude the light e.g:

image

Another way is to enter it manually into the Home Assistant config files: https://www.home-assistant.io/docs/configuration/customizing-devices/#manual-customization

image

This protocol will only do lights - if you want to use it with a Home Assistant `switch` use this Home Assistant integratino to expose the switch as a light: https://www.home-assistant.io/integrations/light.switch/

### Configure Rooms/Zones/Groups in Home Assistant

You can set lights to go into a Room/Zone/Group by customizing entities.

Customize the entity and add the following key and the value is the name you want to use:
- "diyhue_room"
- "diyhue_zone"
- "diyhue_group"

The integration will add it to a room with exactly the same name, if one doesn't exist then it will create one.

You can also set the icon if a room is created by setting `diyhue_class` this has to be a valid icon string otherwise the Hue app will crash. To find valid icon strings set an icon in a room and look in the diyhue app for the group and look at it's "class".

Some valid examples: 

- "Attic"
- "Barbecue"
- "Bathroom"
- "Bedroom"
- "Garden"
- "Hallway"
- "Kids bedroom"
- "Kitchen"
- "Living room"
- "Lounge"
- "Nursery"
- "Staircase"
- "Storage"
- "Studio"
- "Terrace"

Full example of adding a light to a room with an icon in Home Assistant using the customize.yaml

```yaml
light.shower_room_spot_lights
  diyhue: include
  diyhue_class: "Bathroom"
  diyhue_room: "Shower Room"
```
 
### Disabling other diyhue integrations

If you are using the Home Assistant integration you may wish to disable the other integrations so that they don't appear in diyhue directly (e.g. Shelly, Tasmota) if you plan to expose these devices by Home Assistant. To do that modify your diyhue config.json as follows:

```json
    "emulator": {
        "esphome": {
            "enabled": false
        },
        "homeassistant": {
            "enabled": true,
            "homeAssistantIncludeByDefault": false,
            "homeAssistantIp": "your ip",
            "homeAssistantPort": 8123,
            "homeAssistantToken": "your token" 
        },
        "hyperion": {
            "enabled": false
        },
        "mqtt": {
            "discoveryPrefix": "homeassistant",
            "enabled": false,
            "mqttPassword": "",
            "mqttPort": 1883,
            "mqttServer": "mqtt",
            "mqttUser": ""
        },
        "network_scan": {
            "enabled": false
        },
        "sensors": {},
        "shelly": {
            "enabled": false
        },
        "tasmota": {
            "enabled": false
        },
        "yeelight": {
            "enabled": false
        }
    },
```

### CODE REVIEW

File: BridgeEmulator/functions/__init__.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ def nextFreeId(bridge_config, element):
 light_types["LCT001"]["config"] = {"archetype": "sultanbulb","direction": "omnidirectional","function": "mixed","startup": {"configured": True, "mode": "powerfail"}}
 light_types["LCT001"]["capabilities"] = {"certified": True,"control": {"colorgamut": [[0.675,0.322],[0.409,0.518],[0.167,0.04]],"colorgamuttype": "B","ct": {"max": 500,"min": 153},"maxlumen": 600,"mindimlevel": 5000},"streaming": {"proxy": False,"renderer": True}},
 
-light_types["LCT015"] = {"type": "Extended color light", "swversion": "1.46.13_r26312", "manufacturername": "Philips"}
+light_types["LCT015"] = {"type": "Extended color light", "swversion": "1.46.13_r26312", "manufacturername": "no"}
```
Comment (Mevel) [inline BridgeEmulator/functions/__init__.py:32]: Why is the `"manufacturername": "no"`

Comment (foxy82) [reply BridgeEmulator/functions/__init__.py]: typo when your kvm isn't connected to the correct machine - fixed

File: BridgeEmulator/default-config.json [hunk 2]
```diff
@@ -583,6 +583,13 @@
         "mqttPort": 1883,
         "mqttServer": "mqtt",
         "mqttUser": ""
+    },
+    "homeassistant": {
+      "enabled": false,
+      "homeassistantIp": "127.0.0.1",
```
Comment (Thomas-Vos) [inline BridgeEmulator/default-config.json:589]:
```suggestion
      "homeAssistantIp": "127.0.0.1", 
```
Should probably be changed to capital A to be consistent with the other keys below (also change usages of this key in code)

**Review by Thomas-Vos — CHANGES_REQUESTED** [review]

File: BridgeEmulator/default-config.json (hunk 2 above)
Comment (foxy82) [reply BridgeEmulator/default-config.json]: Thanks - fixed.

**Review by Mevel — APPROVED** [review]

**Comment by Shaggy013** [comment]
ehh isnt homeassistant not already with mqtt  

ow and milight is also in home assistant so you aslo need every other light wich both can see .

edit automation.yaml to parse the light to another topic wich you want to control with diyhue 
just like the milight remote example in the ha wiki 

so you dont need make any adjustments then edit a yaml dile

**Comment by mariusmotea** [comment]
@foxy82 can you take a look at the issue reported here https://github.com/diyhue/diyHue/issues/562 ?

Seams the lights where not imported, despite the logs show this.

**Pull request merged**